packages/base/src/panelview/components/filter-panel/Filter.tsx

```diff
@@ -0,0 +1,324 @@
+    this._tracker = options.tracker;
+
+    this.id = 'jupytergis::layerTree';
+    // this.addClass(LAYERS_PANEL_CLASS);
+
+    this.addWidget(
+      ReactWidget.create(
+        <FilterComponent
+          model={this._model}
+          tracker={this._tracker}
+        ></FilterComponent>
+      )
+    );
+  }
+
+  private _model: IControlPanelModel | undefined;
+  private _tracker: IJupyterGISTracker;
+}
+
+interface IFilterComponentProps {
+  model: IControlPanelModel;
+  tracker: IJupyterGISTracker;
+}
+
+const FilterComponent = (props: IFilterComponentProps) => {
+  const featuresInLayerRef = useRef({});
+  const [widgetId, setWidgetId] = useState('');
+  const [logicalOp, setLogicalOp] = useState('all');
+  const [selectedLayer, setSelectedLayer] = useState('');
+  const [filterRows, setFilterRows] = useState<IJGISFilterItem[]>([]);
+  const [featuresInLayer, setFeaturesInLayer] = useState<
+    Record<string, Set<string | number>>
+  >({});
+  const [model, setModel] = useState<IJupyterGISModel | undefined>(
+    props.model.jGISModel
+  );
+
+  props.model?.documentChanged.connect((_, widget) => {
+    setModel(widget?.context.model);
+  });
+
+  // Reset state values when current widget changes
+  useEffect(() => {
+    const handleCurrentChanged = () => {
+      if (props.tracker.currentWidget?.id === widgetId) {
+        return;
+      }
+
+      if (props.tracker.currentWidget) {
+        setWidgetId(props.tracker.currentWidget.id);
+      }
+      setFeaturesInLayer({});
+      setFilterRows([]);
+      setLogicalOp('all');
+      setSelectedLayer('');
+    };
+    props.tracker.currentChanged.connect(handleCurrentChanged);
+
+    return () => {
+      props.tracker.currentChanged.disconnect(handleCurrentChanged);
+    };
+  }, []);
+
+  useEffect(() => {
+    // Keep layer selected when widget changes
+    if (model?.localState?.selected?.value) {
+      setSelectedLayer(Object.keys(model?.localState?.selected?.value)[0]);
+    }
+  }, [widgetId]);
+
+  useEffect(() => {
+    const handleClientStateChanged = () => {
+      if (!model?.localState?.selected?.value) {
+        return;
+      }
+
+      // TODO: handle multi select better
+      const currentLayer = Object.keys(model?.localState?.selected?.value)[0];
+      setSelectedLayer(currentLayer);
+    };
+
+    const handleSharedOptionsChanged = (_, keys) => {
+      // model changes when current widget changes, don't want this to run in that case
+      if (props.tracker.currentWidget?.id === widgetId && keys.has('zoom')) {
+        if (!model?.localState?.selected?.value) {
+          return;
+        }
+        const currentLayer = Object.keys(model?.localState?.selected?.value)[0];
+
+        // TODO: Probably want to debounce/throttle here
+        buildFilterDebounce(currentLayer);
+      }
+    };
+
+    model?.clientStateChanged.connect(handleClientStateChanged);
+
+    // Want to rebuild filter object when zoom changes to get values for that zoom level
```

Comment:
```suggestion
    // Want to rebuild filter object when zoom changes to get values for that zoom level
    // This is because the filtering inputs may depend on the currently visible features
```